contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java

```diff
@@ -0,0 +1,75 @@
+package org.matsim.contrib.drt.optimizer.insertion;
+
+import org.matsim.api.core.v01.Id;
+import org.matsim.contrib.drt.optimizer.VehicleEntry;
+import org.matsim.contrib.drt.optimizer.Waypoint;
+import org.matsim.contrib.drt.passenger.AcceptedDrtRequest;
+import org.matsim.contrib.drt.passenger.DrtRequest;
+import org.matsim.contrib.dvrp.optimizer.Request;
+import org.matsim.contrib.dvrp.passenger.PassengerDroppedOffEvent;
+import org.matsim.contrib.dvrp.passenger.PassengerDroppedOffEventHandler;
+import org.matsim.contrib.dvrp.passenger.PassengerPickedUpEvent;
+import org.matsim.contrib.dvrp.passenger.PassengerPickedUpEventHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MaxDetourInsertionCostCalculator implements InsertionCostCalculator, PassengerPickedUpEventHandler, PassengerDroppedOffEventHandler {
+
+	private final Map<Id<Request>, Double> ongoingRequests = new HashMap<>();
+
+	private final InsertionCostCalculator delegate;
+
+	public MaxDetourInsertionCostCalculator(InsertionCostCalculator delegate) {
+		this.delegate = delegate;
+	}
+
+	@Override
+	public double calculate(DrtRequest drtRequest, InsertionGenerator.Insertion insertion, InsertionDetourTimeCalculator.DetourTimeInfo detourTimeInfo) {
+		if (violatesDetour(insertion, insertion.vehicleEntry, drtRequest)) {
+			return INFEASIBLE_SOLUTION_COST;
+		}
+		return delegate.calculate(drtRequest, insertion, detourTimeInfo);
+	}
+
+	private boolean violatesDetour(InsertionGenerator.Insertion insertion, VehicleEntry vEntry, DrtRequest drtRequest) {
+		// Check if the max travel time constraint for the newly inserted request is violated
+		double rideDuration = insertion.dropoff.newWaypoint.getArrivalTime() - insertion.pickup.newWaypoint.getDepartureTime();
+		if(drtRequest.getMaxRideDuration() < rideDuration) {
+			return true;
+		}
+
+		final int pickupIdx = insertion.pickup.index;
+		Map<Id<Request>, Double> pickUps = new HashMap<>(ongoingRequests);
```

Comment:
What is the point of copying this map?

Reply:
I think I understand the idea. However, I think copying the map and then adding even more stuff to it is a bit too heavy.